src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1084 @@
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(node));
+            }
+
+            private BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern)
+            {
+                return WithInputTypeCheckIfNeeded(pattern, _inputType);
+            }
+
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                // Produce `PatternInputType and pattern` given a new input type
+
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                BoundPattern typePattern = new BoundTypePattern(pattern.Syntax,
+                    new BoundTypeExpression(pattern.Syntax, aliasOpt: null, pattern.InputType),
+                    isExplicitNotNullTest: false, inputType, narrowedType: pattern.InputType);
+
+                var result = new BoundBinaryPattern(pattern.Syntax, disjunction: false, left: typePattern, right: pattern, inputType, pattern.NarrowedType);
+
+                if (pattern.WasCompilerGenerated)
+                {
+                    result = result.MakeCompilerGenerated();
+                }
+
+                return result;
+            }
+
+            public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+            {
+                BoundPattern result;
+                if (node.IsVar)
+                {
+                    result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                }
+                else
+                {
+                    if (node.InputType.Equals(node.DeclaredType.Type, TypeCompareKind.AllIgnoreOptions))
+                    {
+                        result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                    }
+                    else
+                    {
+                        result = node;
+                    }
+                }
+
+                return NegateIfNeeded(WithInputTypeCheckIfNeeded(result));
+            }
+
+            public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
+            {
+                // If we're starting with `Type (D1, D2, ...) { Prop1: P1, Prop2: P2, ... } x`
+                // - if we are not negating, we can expand it to 
+                //   `Type (D1, _, ...) and Type (_, D2, ...) and Type { Prop1: P1 } and Type { Prop2: P2 } ...` 
```

Comment:
>`Type (D1, _, ...) and Type (_, D2, ...) and Type { Prop1: P1 } and Type { Prop2: P2 } ...` [](http://example.com/codeflow?start=21&length=91)

Should we add a stand-alone type check at the beginning?

Reply:
I don't have a good/principled way to answer that. I'm trying the change and the only effect on existing tests is that why syntax is reported as redundant.